### PR TITLE
Conda build

### LINKED ISSUE #37584: Add a CI job using conda compilers
## 🐛 Bug

There's regularly issues building PyTorch with either conda compilers or with a very new (system installed) GCC + conda Python+dependencies. Using a development setup based on conda is recommended (https://github.com/pytorch/pytorch#from-source) so this is not a great situation. Example issues:

- https://github.com/pytorch/pytorch/issues/29093 (has some relevant discussion)
- https://github.com/pytorch/cpuinfo/issues/36
- The patches we're carrying around for a build setup used by ~8 people, gets quite inefficient: https://github.com/Quansight/pearu-sandbox/blob/master/working-envs/activate-pytorch-dev.sh

Furthermore, making the build config with conda compilers robust is required for making the conda binaries conda-forge compatible so they can be synced to `conda-forge/pytorch` (there's more to do there that I need to follow up on, this is step 1 - xref https://github.com/conda-forge/pytorch-cpu-feedstock/issues/7).

## Expected behavior

Using conda compilers on Linux should work out of the box. To make it more reliable, we should add a CI job using it. The question I have is if it should be an extra CI job entry, or if we should modify an existing one (and if so, which one)?

Cc @ezyang 

cc @malfet @ezyang @seemethere

### PR DESCRIPTION
**Pull request by mattip**
closes gh-37584. ~I think I need to do more to generate an image, but the `.circleci/README.md` is vague in the details. The first commit reflows and updates that document a bit, I will continue to update it as the PR progresses :)~ Dropped updating `.circleci/README.md`, will do that in a separate PR once this is merged.

### CODE REVIEW

File: .circleci/docker/common/install_conda.sh
Comment (malfet) [inline .circleci/docker/common/install_conda.sh:74]: I'm about to land this change as part of https://github.com/pytorch/pytorch/pull/38747

Comment (mattip) [reply .circleci/docker/common/install_conda.sh:74]: Thanks for the heads up. How do I get the new tag for the `DOCKER_IMAGE_TAG`?

Comment (malfet) [reply .circleci/docker/common/install_conda.sh:74]: It's a hash sum of CircleCI setup job on your PR. Please make sure to add new image to https://github.com/pytorch/pytorch/blob/master/.circleci/verbatim-sources/workflows/workflows-ecr-gc.yml#L13

And to trigger build on PR, delete `triggers:` subsection from https://github.com/pytorch/pytorch/blob/master/.circleci/verbatim-sources/workflows/workflows-docker-builder.yml#L2

Comment (mattip) [reply .circleci/docker/common/install_conda.sh:74]: > It's a hash sum of CircleCI setup job on your PR. Please make sure to add ...

How do I find/calculate the hash sum?

**Comment by mattip** [comment]
@malfet I assume at some point I will need to update the `cimodel` with the information that `conda` is a valid compiler specifier?

File: .circleci/docker/common/install_conda.sh
Comment (mattip) [reply .circleci/docker/common/install_conda.sh:74]: @malfet: how do I find/calculate the hash sum?

Comment (mattip) [reply .circleci/docker/common/install_conda.sh:74]: @malfet I removed the `triggers:` section and now see a few more `ecr_gc_job_for*` CI jobs, but none of them seem to be docker builds. What am I missing?

**Comment by mattip** [comment]
Rebased off master to clear the merge conflicts. @malfet I still don't understand how to get the proper hash to put into the ignore stanza, and it seems the actual docker image build is not being triggered.

File: .circleci/docker/common/install_conda.sh
Comment (mattip) [reply .circleci/docker/common/install_conda.sh]: There are hints in gh-38686, it seems the checksum is `DOCKER_TAG=$(git rev-parse HEAD:.circleci/docker)`. But perhaps this PR should wait for that one which will make all this easier.

Comment (mattip) [reply .circleci/docker/common/install_conda.sh]: it seems not so simple. `git rev-parse HEAD:.circleci/docker` gives a hash as a string, for instance `0b8f43b056e9b40f7b7b0747883d200e36fcd227`, but the needed tag is of the format `209062ef-ab58-422a-b295-36c4eed6e906`

**Comment by mattip** [comment]
ping @malfet

**Comment by mattip** [comment]
@seemethere is there a way to move this forward? I am stuck trying to generate a new Docker image for the CI run

**Comment by mattip** [comment]
This PR is still stuck waiting for instructions on how to generate a new CI docker image.

**Comment by mattip** [comment]
@ezyang is there something I can do to move this forward?

**Comment by ezyang** [comment]
Hi @mattip, a lot of us are busy right now because it is performance cycle at FB.

I don't have too much context on this diff, but if I understand correctly, on master, it's no longer necessary to manually delete the trigger strings, as this diff from @seemethere has landed: #40194. This means that the instructions at https://github.com/pytorch/pytorch/wiki/Docker-image-build-on-CircleCI are out of date.

All you have to do now is submit a PR with the new Docker changes, and CI will automatically build the new images and test against them, and if all is green and this PR is approved I can land it.

**Comment by mattip** [comment]
> All you have to do now is submit a PR with the new Docker changes, and CI will automatically build the new images and test against them, and if all is green and this PR is approved I can land it.

I tried to add a  `pytorch-linux-bionic-py3.7-conda` image, but I do not see them among the circleCI runs.

**Comment by ezyang** [comment]
I read over the CircleCI yaml, and it looks like I was mistaken and it's still necessary to delete the triggers, sorry :( I'm going to try pushing the necessary update to trigger the build. Also, it looks like you don't have the ECR repo setup so I went ahead and did that for you. 

For future reference, try to avoid force pushing to the branch, it destroys history on GitHub which means I can't look at old builds and see if you previously had things setup correctly.

**Comment by ezyang** [comment]
docker build: https://app.circleci.com/pipelines/github/pytorch/pytorch/191859/workflows/c966facd-bcb3-4b37-ba7d-f9b995ad78d9

**Comment by mattip** [comment]
> try to avoid force pushing to the branch, it destroys history on GitHub

Sorry. So the preferred workflow is to merge master into the branch?

**Comment by ezyang** [comment]
Yes, just merge in master. When we land the diff we'll squash all history so the merges won't matter in the long run.

**Comment by mattip** [comment]
The image has been built and [uploaded](http://docker.pytorch.org/pytorch.html), so I added it to the data model in 14b8f86

**Comment by mattip** [comment]
It seems to work, the test failures are the same as the other builds.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
The test failures look problematic, they don't seem to be due to master problem:

```

Jul 20 08:04:54 ======================================================================
Jul 20 08:04:54 FAIL [0.177s]: test_figure (__main__.TestTensorBoardFigure)
Jul 20 08:04:54 ----------------------------------------------------------------------
Jul 20 08:04:54 Traceback (most recent call last):
Jul 20 08:04:54   File "test_tensorboard.py", line 642, in test_figure
Jul 20 08:04:54     self.assertFalse(plt.fignum_exists(figure.number))
Jul 20 08:04:54 AssertionError: True is not false
Jul 20 08:04:54 
Jul 20 08:04:54 ======================================================================
Jul 20 08:04:54 FAIL [1.110s]: test_figure_list (__main__.TestTensorBoardFigure)
Jul 20 08:04:54 ----------------------------------------------------------------------
Jul 20 08:04:54 Traceback (most recent call last):
Jul 20 08:04:54   File "test_tensorboard.py", line 664, in test_figure_list
Jul 20 08:04:54     self.assertTrue(all([plt.fignum_exists(figure.number) is False for figure in figures]))  # noqa F812
Jul 20 08:04:54 AssertionError: False is not true
Jul 20 08:04:54 
Jul 20 08:04:54 ----------------------------------------------------------------------
Jul 20 08:04:54 Ran 50 tests in 41.752s
```

it looks like there might be some sort of tensorboard update that broke the tests. May need to pin.

**Comment by ezyang** [comment]
I think it might be a matplotlib problem

**Comment by ezyang** [comment]
Redoing build with matplotlib pin

**Comment by ezyang** [comment]
Docker build: https://app.circleci.com/pipelines/github/pytorch/pytorch/191941/workflows/d368bea1-af47-48fd-a3b7-c41d1e7b4f1a

**Comment by mattip** [comment]
I updated the tag from `c966facd-bcb3-4b37-ba7d-f9b995ad78d9` to `8bdba785b1eac4d297d5f5930f979518012a56e0` which is what appears on http://docker.pytorch.org/pytorch.html. I hope that is correct, the format is different.

**Comment by mattip** [comment]
The lint quickcheck of `.circleci/config.yml` failed, but I fail to see why. There is no change if I call the rengenerate.sh script locally.

**Comment by ezyang** [comment]
8bdba785b1eac4d297d5f5930f979518012a56e0 looks wrong, there is no pytorch-linux-bionic-py3.7-conda build. You want d368bea1-af47-48fd-a3b7-c41d1e7b4f1a

**Comment by mattip** [comment]
The `pytorch_linux_xenial_py3_clang5_android_ndk_r19c` CI runs are [failing to find an image](https://app.circleci.com/pipelines/github/pytorch/pytorch/192063/workflows/7a8d698c-1b6d-4a54-a336-f6e0a7598907/jobs/6301790) with the hash `d368bea1-af47-48fd-a3b7-c41d1e7b4f1a`. All the other images _are_ finding that hash. When I look at the [docker build CI run](https://app.circleci.com/pipelines/github/pytorch/pytorch/191941/workflows/d368bea1-af47-48fd-a3b7-c41d1e7b4f1a), I see that the other images succeeded but that one is still marked as failed. I have re-triggered it  via the "rerun job with ssh" button, but then the job runs successfully claiming that the image hash `8bdba785b1eac4d297d5f5930f979518012a56e0` already exits. And indeed, if I go to the [docker image listing](http://docker.pytorch.org/pytorch.html), I see an image for `pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c` with that hash.

When I look at some of the other successful docker builds, the "Check if image should be built" step also looks for `8bdba7...`. On the docker image listing, I see both hashes: the `8bdba...` one and the `d368bea1-af47...` one. What am I missing?

**Comment by ezyang** [comment]
> I have re-triggered it via the "rerun job with ssh" button, but then the job runs successfully claiming that the image hash 8bdba785b1eac4d297d5f5930f979518012a56e0 already exits. And indeed, if I go to the docker image listing, I see an image for pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c with that hash.

There's a bug; the docker jobs are checking for the wrong hash. The bug was exacerbated by the fact that streamlined docker build workflow was only partially landed (and part of it was reverted.) I think we will need to work around the bug for now, probably by deleting the "should build image" test. cc @seemethere I don't have time to fix this today due to PSC but will get back on it tomorrow.

**Comment by ezyang** [comment]
OK, turns out I was wrong.

> 8bdba785b1eac4d297d5f5930f979518012a56e0 looks wrong, there is no pytorch-linux-bionic-py3.7-conda build. You want d368bea1-af47-48fd-a3b7-c41d1e7b4f1a

This is wrong: the image does exist, as per:

```
(/scratch/ezyang/pytorch-scratch2-env) ezyang@devfair040:/scratch/ezyang$ docker run -it 308535385114.dkr.e
cr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-py3.7-conda:8bdba785b1eac4d297d5f5930f979518012a56e
0
Unable to find image '308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-py3.7-conda
:8bdba785b1eac4d297d5f5930f979518012a56e0' locally
8bdba785b1eac4d297d5f5930f979518012a56e0: Pulling from pytorch/pytorch-linux-bionic-py3.7-conda
```

It isn't showing up in the index, probably because the index hasn't picked up the new ECR repo. So actually you were all fine.

The real reason why quickcheck lint was failing, was because there was an update with master, and the quickcheck is computed after merging with master (kind of icky, but you would have indeed broken quickcheck if you had merged as is. Would have been nice if it mentioned this though.)

I'm going to revert back to your diffs and try to land.

**Comment by mattip** [comment]
Thanks. I still think think it is fragile to tag the images with the `CIRCLE_WORKFLOW_ID`, but hopefully that is orthogonal to this PR.

**Comment by ezyang** [comment]
I think in the half-constructed state we are in today, you basically can't use workflow ids, for the reasons you posted above.

**Comment by jakirkham** [comment]
Woot! 🎉 Thanks Matti! 😄

**Comment by rgommers** [comment]
This PR was reverted, with reason not documented and no further follow-up as far as I can tell. And right now there's again no CI job with conda compilers, and the build seems to be broken for at least some people (gh-47717).

@seemethere do you remember what happened here? Can we re-introduce a conda compiler CI job?

**Pull request closed without merging**
